### DISCUSSION TITLE
How to use Poetry with PyCharm? All imports show warning `No module named...`.

### DISCUSSION THREAD

**Discussion by simon-liebehenschel** [python-poetry/poetry#5119] [Q&A] [answered]
- [x] I have "pyproject.toml" and "poetry.lock" files.
- [x] I have all packages installed via `poetry install`
- [x] **I can run my code that uses installed packages.**

 PyCharm 2021.3.1

# Problem

When I import any package in PyCharm, I see  `No module named 'blablabla' `. 

# Question

How to support packages imports, auto-completion? 

How to view packages source code?

# Workaround

I have to install all packages via "Pip" in parallel to hide PyCharm import warnings `No module named 'blablabla' `. This is super stupid.

**Comment by jacob-lee** [comment]
Pycharm 2021.3 now offers Poetry support. Is that the version of Pycharm you are using?

**Reply by simon-liebehenschel** [reply]
I already wrote PyCharm version in my post. Read my post again. Yes, the version is "21.3.2".

**Reply by jacob-lee** [reply]
Apologies.

**Reply by jacob-lee** [reply]
I recommend posting this to [JetBrains issue tracker](https://youtrack.jetbrains.com/issues?q=pycharm%20poetry), if you haven't done so already.

**Comment by emilhe** [comment]
Did you select the poetry environment as interpreter? After I did that, everything works as expected.

**Reply by simon-liebehenschel** [reply]
Of course I selected poetry environment as interpreter.

PyCharm is a full of bugs software, so probably it does not work well when a pip was used before Poetry for the same project.

**Comment by simon-liebehenschel** [comment, marked as answer]
# Solution:
1. Uninstall Pycharm
2. Remove your project
3. Remove cache files

Install PyCharm again, add the project again, add an interpreter again.

Now imports are not red. Now I can autocomplete imports.

**Reply by elliot-100** [reply]
Hi, having similar issue.

- What do you mean by "Remove your project" given you've already uninstalled Pycharm at this point? e.g. delete all the files?
- "Remove cache files" - Pycharm's or Poetry's?

Thanks!

**Reply by simon-liebehenschel** [reply]
@elliot-100 Sadly, the solution above is not persistent. If you want to make PyCharm work well with Poetry, then I recommend to use [`virtualenvs.in-project = true`](https://python-poetry.org/docs/configuration/#virtualenvsin-project) to create venv within the root directory of the project. This way PyCharm config for a project will not disappear randomly on next PyCharm restart.

---

By the way, I migrated from PyCharm because it [indexing (scanning) a whole .venv on every restart](https://intellij-support.jetbrains.com/hc/en-us/community/posts/206246539-how-to-disable-indexing-on-startup) which is extremely slow... perhaps PyCharm developers do not know about caching. I will not write to what I switched to avoid someone thinking that I am advertising here another IDEs, but I highly recommend to give a try with a few other popular IDE if you work with Python.

**Reply by elliot-100** [reply]
Thanks, I'll try putting the virtualenvs in the project.

Yes, am also finding Pycharm frustrating at the moment. Am only working on small projects, so reindexing isn't too much of an issue. But its Poetry (and other environments) handling seems very unreliable.
